File: study_tracker_flutter/lib/data/repositories/session_repository.dart
import 'package:drift/drift.dart';
import '../db/database.dart' as db;
import '../db/database_provider.dart';
import '../../domain/domain.dart' as domain;
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../core/sync/sync_trigger.dart';


/// Session repository — all CRUD and query operations for study_sessions.
class SessionRepository {
  final db.AppDatabase _db;
  List<domain.StudySession>? _allSessionsCache;
  domain.SessionSummary? _summaryCache;
  final Map<int, List<domain.StudySession>> _recentCache = {};
  int? _cachedProfileId;

  SessionRepository(this._db);

  Future<int> _activeProfileId() async {
    final row = await (_db.select(_db.appSettings)
          ..where((t) => t.key.equals('currentProfileId')))
        .getSingleOrNull();
    final profileId = int.tryParse(row?.value ?? '1') ?? 1;
    if (_cachedProfileId != profileId) {
      _cachedProfileId = profileId;
      _invalidateCaches();
    }
    return profileId;
  }

  void _invalidateCaches() {
    _allSessionsCache = null;
    _summaryCache = null;
    _recentCache.clear();
  }

  void clearLocalCache() {
    _invalidateCaches();
  }

  // ── Validation ──────────────────────────────────────────────────

  void _validate(domain.StudySession session) {
    if (session.startAt.isAfter(session.endAt)) {
      throw ArgumentError('startAt must be before endAt.');
    }
    if (session.durationMinutes <= 0) {
      throw ArgumentError('durationMinutes must be positive.');
    }
  }

  // ── Create ──────────────────────────────────────────────────────

  Future<int> create(domain.StudySession session) async {
    _validate(session);
    final profileId = await _activeProfileId();
    final id = await _db.into(_db.studySessions).insert(
          db.StudySessionsCompanion.insert(
            profileId: Value(profileId),
            startAt: session.startAt.toIso8601String(),
            endAt: session.endAt.toIso8601String(),
            durationMinutes: session.durationMinutes,
            subjectId: Value(session.subjectId),
            subjectName: Value(session.subjectName),
            topic: Value(session.topic),
            chapterTag: Value(session.chapterTag),
            mood: Value(session.mood),
            notes: Value(session.notes),
            mode: session.mode.dbValue,
            breakMinutes: Value(session.breakMinutes),
            backgroundImage: Value(session.backgroundImage),
            updatedAt: Value(DateTime.now()),
          ),
        );
    // Insert tasks if any
    for (final task in session.tasks) {
      await _db.into(_db.sessionTasks).insert(
            db.SessionTasksCompanion.insert(
              profileId: Value(profileId),
              sessionId: id,
              title: task.title,
              completed: Value(task.completed),
            ),
          );
    }
    _invalidateCaches();
    SyncTrigger.instance.notifyWrite();
    return id;

  }

  // ── Update ──────────────────────────────────────────────────────

  Future<void> update(domain.StudySession session) async {
    if (session.id == null) throw ArgumentError('Session id required for update.');
    _validate(session);
        final profileId = await _activeProfileId();
    await (_db.update(_db.studySessions)
          ..where((t) => t.id.equals(session.id!) & t.profileId.equals(profileId)))
        .write(db.StudySessionsCompanion(
          startAt: Value(session.startAt.toIso8601String()),
          endAt: Value(session.endAt.toIso8601String()),
      durationMinutes: Value(session.durationMinutes),
      subjectId: Value(session.subjectId),
      subjectName: Value(session.subjectName),
      topic: Value(session.topic),
      chapterTag: Value(session.chapterTag),
      mood: Value(session.mood),
      notes: Value(session.notes),
      mode: Value(session.mode.dbValue),
      breakMinutes: Value(session.breakMinutes),
      backgroundImage: Value(session.backgroundImage),
      updatedAt: Value(DateTime.now()),
    ));
    _invalidateCaches();
    SyncTrigger.instance.notifyWrite();
  }


  // ── Delete ──────────────────────────────────────────────────────

  Future<void> delete(int id) async {
    final profileId = await _activeProfileId();
    await (_db.update(_db.studySessions)
          ..where((t) => t.id.equals(id) & t.profileId.equals(profileId)))
        .write(db.StudySessionsCompanion(
          deletedAt: Value(DateTime.now().toUtc().toIso8601String()),
          updatedAt: Value(DateTime.now()),
        ));
    _invalidateCaches();
    SyncTrigger.instance.notifyWrite();
  }


  // ── Read ────────────────────────────────────────────────────────

  Future<domain.StudySession?> getById(int id) async {
    final profileId = await _activeProfileId();
    final row = await (_db.select(_db.studySessions)
          ..where((t) => t.id.equals(id) & t.profileId.equals(profileId) & t.deletedAt.isNull()))
        .getSingleOrNull();
    if (row == null) return null;
    final tasks = await _getTasksForSession(id);
    return _mapRow(row, tasks);
  }

  /// Fetch sessions with optional filter.
  Future<List<domain.StudySession>> list(domain.SessionFilter filter) async {
    final profileId = await _activeProfileId();
    final isUnfiltered = filter.subjectId == null &&
        filter.groupId == null &&
        filter.mode == null &&
        filter.startFrom == null &&
        filter.startTo == null &&
        (filter.query == null || filter.query!.isEmpty) &&
        filter.offset == 0;

    if (isUnfiltered) {
      if (filter.limit >= 9999 && _allSessionsCache != null) {
        return _allSessionsCache!;
      }
      final recentCached = _recentCache[filter.limit];
      if (recentCached != null) {
        return recentCached;
      }
    }

    List<int>? groupSubjectIds;
    if (filter.groupId != null) {
      final groupRows = await (_db.select(_db.subjects)
        ..where((s) => s.groupId.equals(filter.groupId!) & s.profileId.equals(profileId)))
          .get();
      groupSubjectIds = groupRows.map((e) => e.id).toList();
      if (groupSubjectIds.isEmpty) {
        return const [];
      }
    }

    var query = _db.select(_db.studySessions);
    query = query
      ..where((t) {
        Expression<bool> expr = const Constant(true);
        expr = expr & t.profileId.equals(profileId) & t.deletedAt.isNull();
        if (filter.subjectId != null) {
          expr = expr & t.subjectId.equals(filter.subjectId!);
        }
        if (groupSubjectIds != null) {
          expr = expr & t.subjectId.isIn(groupSubjectIds);
        }
        if (filter.mode != null) {
          expr = expr & t.mode.equals(filter.mode!.dbValue);
        }
        if (filter.startFrom != null) {
          expr = expr &
              t.startAt.isBiggerOrEqualValue(filter.startFrom!.toIso8601String());
        }
        if (filter.startTo != null) {
          expr = expr &
              t.startAt.isSmallerOrEqualValue(filter.startTo!.toIso8601String());
        }
        if (filter.query != null && filter.query!.isNotEmpty) {
          final q = '%${filter.query}%';
          expr = expr &
              (t.subjectName.like(q) | t.topic.like(q) | t.notes.like(q));
        }
        return expr;
      })
      ..orderBy([(t) => OrderingTerm.desc(t.startAt)])
      ..limit(filter.limit, offset: filter.offset);

    final rows = await query.get();
    final mapped = rows.map((r) => _mapRow(r, const [])).toList();

    if (isUnfiltered) {
      if (filter.limit >= 9999) {
        _allSessionsCache = mapped;
      } else {
        _recentCache[filter.limit] = mapped;
      }
    }

    return mapped;
  }

    Future<List<domain.StudySession>> getRecent(int limit) =>
      list(domain.SessionFilter(limit: limit));

  /// Summary stats — total sessions, total minutes, average, recent.
  Future<domain.SessionSummary> getSummary() async {
    final profileId = await _activeProfileId();
    if (_summaryCache != null) return _summaryCache!;

    final result = await _db.customSelect(
      'SELECT COUNT(*) AS total_sessions, '
      'COALESCE(SUM(duration_minutes), 0) AS total_minutes, '
      'COALESCE(AVG(duration_minutes), 0) AS average_minutes '
      'FROM study_sessions WHERE profile_id = ? AND deleted_at IS NULL',
      variables: [Variable.withInt(profileId)],
    ).getSingle();
    final recent = await getRecent(5);
    final summary = domain.SessionSummary(
      totalSessions: result.data['total_sessions'] as int,
      totalMinutes: result.data['total_minutes'] as int,
      averageMinutes: (result.data['average_minutes'] as num).toDouble(),
      recentSessions: recent,
    );
    _summaryCache = summary;
    return summary;
  }

  /// Summary stats for the last 7 days only.
  Future<domain.SessionSummary> getSummaryLast7Days() async {
    final profileId = await _activeProfileId();
    final now = DateTime.now();
    final cutoffDate = DateTime(now.year, now.month, now.day).subtract(const Duration(days: 7));
    final cutoff = cutoffDate.toIso8601String();
    final result = await _db.customSelect(
      'SELECT COUNT(*) AS total_sessions, '
      'COALESCE(SUM(duration_minutes), 0) AS total_minutes, '
      'COALESCE(AVG(duration_minutes), 0) AS average_minutes '
      'FROM study_sessions '
      'WHERE profile_id = ? AND deleted_at IS NULL AND start_at >= ?',
      variables: [Variable<int>(profileId), Variable<String>(cutoff)],
    ).getSingle();
    final recent = await list(domain.SessionFilter(limit: 5, startFrom: cutoffDate));
    return domain.SessionSummary(
      totalSessions: result.data['total_sessions'] as int,
      totalMinutes: result.data['total_minutes'] as int,
      averageMinutes: (result.data['average_minutes'] as num).toDouble(),
      recentSessions: recent,
    );
  }

  /// Current study streak (consecutive days with sessions).
  Future<int> getCurrentStreak({DateTime? now}) async {
    final profileId = await _activeProfileId();
    final effectiveNow = now ?? DateTime.now();
    final cutoff = effectiveNow.subtract(const Duration(days: 90)).toIso8601String();
    final rows = await _db.customSelect(
      'SELECT date(start_at) AS study_day '
      'FROM study_sessions '
      'WHERE profile_id = ? AND deleted_at IS NULL AND start_at >= ? '
      'GROUP BY study_day '
      'ORDER BY study_day DESC',
      variables: [Variable<int>(profileId), Variable<String>(cutoff)],
    ).get();

    if (rows.isEmpty) return 0;

    final days = rows
        .map((row) => DateTime.parse(row.data['study_day'] as String))
        .toList();

    DateTime cursor = DateTime(effectiveNow.year, effectiveNow.month, effectiveNow.day);
    final hasToday = days.any((d) => _isSameDay(d, cursor));
    if (!hasToday) cursor = cursor.subtract(const Duration(days: 1));

    var streak = 0;
    while (days.any((d) => _isSameDay(d, cursor))) {
      streak += 1;
      cursor = cursor.subtract(const Duration(days: 1));
    }

    return streak;
  }

  bool _isSameDay(DateTime a, DateTime b) {
    return a.year == b.year && a.month == b.month && a.day == b.day;
  }

  /// Most-studied subject over the last 7 days (by total minutes).
  Future<String?> getMostStudiedSubjectLast7Days() async {
    final profileId = await _activeProfileId();
    final now = DateTime.now();
    final cutoff = DateTime(now.year, now.month, now.day).subtract(const Duration(days: 7)).toIso8601String();
    final result = await _db.customSelect(
      'SELECT subject_name, COALESCE(SUM(duration_minutes), 0) AS total_minutes '
      'FROM study_sessions '
      'WHERE profile_id = ? AND deleted_at IS NULL '
      'AND subject_name IS NOT NULL AND subject_name != \'\' '
      'AND start_at >= ? '
      'GROUP BY subject_name '
      'ORDER BY total_minutes DESC '
      'LIMIT 1',
      variables: [Variable<int>(profileId), Variable<String>(cutoff)],
    ).getSingleOrNull();

    return result?.data['subject_name'] as String?;
  }

  /// Total study minutes from Monday 00:00 through now (for goal progress calculation).
  Future<int> getTotalMinutesThisWeek() async {
    final profileId = await _activeProfileId();
    final now = DateTime.now();
    // Monday = 1, Sunday = 7; calculate Monday of this week at 00:00
    final dayOfWeek = now.weekday; // 1=Mon, 7=Sun
    final mondayThisWeek = now.subtract(Duration(days: dayOfWeek - 1));
    final mondayStart = DateTime(mondayThisWeek.year, mondayThisWeek.month, mondayThisWeek.day);
    final cutoff = mondayStart.toIso8601String();
    
    final result = await _db.customSelect(
      'SELECT COALESCE(SUM(duration_minutes), 0) AS total_minutes '
      'FROM study_sessions '
      'WHERE profile_id = ? AND deleted_at IS NULL AND start_at >= ?',
      variables: [Variable<int>(profileId), Variable<String>(cutoff)],
    ).getSingle();
    
    return result.data['total_minutes'] as int;
  }

  /// Most recent session's start date/time, or null if no sessions.
  Future<DateTime?> getLastSessionDate() async {
    final profileId = await _activeProfileId();
    final result = await _db.customSelect(
      'SELECT start_at FROM study_sessions '
      'WHERE profile_id = ? AND deleted_at IS NULL '
      'ORDER BY start_at DESC LIMIT 1',
      variables: [Variable<int>(profileId)],
    ).getSingleOrNull();
    
    if (result == null) return null;
    return DateTime.parse(result.data['start_at'] as String);
  }

  /// Elapsed days in current week (1 = Monday, 7 = Sunday).
  /// Used for goal progress pace calculation.
  int getElapsedDaysThisWeek({DateTime? now}) {
    final effectiveNow = now ?? DateTime.now();
    // weekday: 1=Mon, 7=Sun. We count Mon as 1, Sun as 7.
    return effectiveNow.weekday;
  }

  Stream<List<domain.StudySession>> watchAllSessions() {
    final profile = _db.select(_db.appSettings)..where((t) => t.key.equals('currentProfileId'));
    return profile
        .watchSingleOrNull()
        .map((row) => int.tryParse(row?.value ?? '1') ?? 1)
        .distinct()
        .asyncExpand((profileId) {
      final query = _db.select(_db.studySessions)
        ..where((t) => t.profileId.equals(profileId) & t.deletedAt.isNull())
        ..orderBy([(t) => OrderingTerm.desc(t.startAt)]);
      return query.watch().map((rows) => rows.map((r) => _mapRow(r, const [])).toList());
    });
  }

  Stream<List<domain.StudySession>> watchList(domain.SessionFilter filter) {
    final profile = _db.select(_db.appSettings)..where((t) => t.key.equals('currentProfileId'));
    return profile
        .watchSingleOrNull()
        .map((row) => int.tryParse(row?.value ?? '1') ?? 1)
        .distinct()
        .asyncExpand((profileId) {
      var query = _db.select(_db.studySessions);
      query = query
        ..where((t) {
          Expression<bool> expr = const Constant(true);
          expr = expr & t.profileId.equals(profileId) & t.deletedAt.isNull();
          if (filter.subjectId != null) {
            expr = expr & t.subjectId.equals(filter.subjectId!);
          }
          if (filter.mode != null) {
            expr = expr & t.mode.equals(filter.mode!.dbValue);
          }
          if (filter.startFrom != null) {
            expr = expr & t.startAt.isBiggerOrEqualValue(filter.startFrom!.toIso8601String());
          }
          if (filter.startTo != null) {
            expr = expr & t.startAt.isSmallerOrEqualValue(filter.startTo!.toIso8601String());
          }
          if (filter.query != null && filter.query!.isNotEmpty) {
            final q = '%${filter.query}%';
            expr = expr & (t.subjectName.like(q) | t.topic.like(q) | t.notes.like(q));
          }
          return expr;
        })
        ..orderBy([(t) => OrderingTerm.desc(t.startAt)])
        ..limit(filter.limit, offset: filter.offset);

      return query.watch().map((rows) => rows.map((r) => _mapRow(r, const [])).toList());
    });
  }

  Stream<List<domain.StudySession>> watchRecentSessions(int limit) {
    final profile = _db.select(_db.appSettings)..where((t) => t.key.equals('currentProfileId'));
    return profile
        .watchSingleOrNull()
        .map((row) => int.tryParse(row?.value ?? '1') ?? 1)
        .distinct()
        .asyncExpand((profileId) {
      final query = _db.select(_db.studySessions)
        ..where((t) => t.profileId.equals(profileId) & t.deletedAt.isNull())
        ..orderBy([(t) => OrderingTerm.desc(t.startAt)])
        ..limit(limit);
      return query.watch().map((rows) => rows.map((r) => _mapRow(r, const [])).toList());
    });
  }

  Stream<domain.SessionSummary> watchTodaySummary() {
    return watchAllSessions().map((sessions) {
      final totalSessions = sessions.length;
      final totalMinutes = sessions.fold<int>(0, (sum, s) => sum + s.durationMinutes);
      final averageMinutes = totalSessions == 0 ? 0.0 : totalMinutes / totalSessions;
      final recentSessions = sessions.take(5).toList();

      return domain.SessionSummary(
        totalSessions: totalSessions,
        totalMinutes: totalMinutes,
        averageMinutes: averageMinutes,
        recentSessions: recentSessions,
      );
    });
  }

  /// All sessions for analytics (no limit).
    Future<List<domain.StudySession>> getAll() =>
      list(const domain.SessionFilter(limit: 9999));

  // ── Helpers ─────────────────────────────────────────────────────

  Future<List<domain.SessionTask>> _getTasksForSession(int sessionId) async {
    final profileId = await _activeProfileId();
    final rows = await (_db.select(_db.sessionTasks)
          ..where((t) => t.sessionId.equals(sessionId) & t.profileId.equals(profileId)))
        .get();
    return rows.map(_mapTaskRow).toList();
  }

  /// Batch-load tasks for many sessions (e.g. AI analytics over a window).
  Future<Map<int, List<domain.SessionTask>>> tasksBySessionIds(Iterable<int> sessionIds) async {
    final ids = sessionIds.toSet().toList();
    if (ids.isEmpty) return {};
    final profileId = await _activeProfileId();
    final rows = await (_db.select(_db.sessionTasks)
          ..where((t) => t.profileId.equals(profileId) & t.sessionId.isIn(ids)))
        .get();
    final map = <int, List<domain.SessionTask>>{};
    for (final r in rows) {
      map.putIfAbsent(r.sessionId, () => []).add(_mapTaskRow(r));
    }
    return map;
  }

  domain.StudySession _mapRow(db.StudySession row, List<domain.SessionTask> tasks) {
    return domain.StudySession(
      id: row.id,
      startAt: DateTime.parse(row.startAt),
      endAt: DateTime.parse(row.endAt),
      durationMinutes: row.durationMinutes,
      subjectId: row.subjectId,
      subjectName: row.subjectName,
      topic: row.topic,
      chapterTag: row.chapterTag,
      mood: row.mood,
      notes: row.notes,
      mode: domain.StudySessionMode.fromDb(row.mode),
      breakMinutes: row.breakMinutes,
      backgroundImage: row.backgroundImage,
      createdAt: row.createdAt,
      updatedAt: row.updatedAt,
      tasks: tasks,
    );
  }

  domain.SessionTask _mapTaskRow(db.SessionTask row) {
    return domain.SessionTask(
      id: row.id,
      sessionId: row.sessionId,
      title: row.title,
      completed: row.completed,
      createdAt: row.createdAt,
    );
  }
}

final sessionRepositoryProvider = Provider<SessionRepository>(
  (ref) => SessionRepository(ref.watch(databaseProvider)),
);

File: src/core/data/repositories/sessionRepository.ts
import { getDatabase } from '../database';
import type { SessionFilter, SessionSummary, StudySession } from '../../domain';
import { normalizeNullableText, normalizeText } from '../sql';
import { getCurrentProfileId } from './profileRepository';
import { syncTrigger } from '../../sync/syncTrigger';

type SessionRow = {
  id: number;
  start_at: string;
  end_at: string;
  duration_minutes: number;
  subject_id: number | null;
  subject_name: string | null;
  topic: string | null;
  chapter_tag: string | null;
  mood: string | null;
  notes: string | null;
  mode: StudySession['mode'];
  break_minutes: number;
  created_at: string;
  updated_at: string;
};

function validateSession(session: StudySession): StudySession {
  const startAt = normalizeText(session.startAt ?? '');
  const endAt = normalizeText(session.endAt ?? '');

  if (!startAt || !endAt) {
    throw new Error('Session start and end times are required.');
  }

  if (session.durationMinutes <= 0) {
    throw new Error('Session duration must be positive.');
  }

  if (!session.mode) {
    throw new Error('Session mode is required.');
  }

  return {
    ...session,
    startAt,
    endAt,
    subjectName: normalizeNullableText(session.subjectName),
    topic: normalizeNullableText(session.topic),
    chapterTag: normalizeNullableText(session.chapterTag),
    mood: normalizeNullableText(session.mood),
    notes: normalizeNullableText(session.notes)
  };
}

export async function createSession(session: StudySession): Promise<number> {
  const normalized = validateSession(session);
  const database = await getDatabase();
  const profileId = await getCurrentProfileId();

  const result = await database.execute(
    `INSERT INTO study_sessions (
      profile_id,
      start_at, end_at, duration_minutes, subject_id, subject_name, topic, chapter_tag,
      mood, notes, mode, break_minutes, sync_id, created_at, updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, lower(hex(randomblob(16))), CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);`,
    [
      profileId,
      normalized.startAt,
      normalized.endAt,
      normalized.durationMinutes,
      normalized.subjectId ?? null,
      normalized.subjectName ?? null,
      normalized.topic ?? null,
      normalized.chapterTag ?? null,
      normalized.mood ?? null,
      normalized.notes ?? null,
      normalized.mode,
      normalized.breakMinutes ?? 0
    ]
  );

  syncTrigger.notifyChange();
  return Number(result.lastInsertId);
}

export async function updateSession(session: StudySession): Promise<void> {
  if (!session.id) {
    throw new Error('Session id is required for updates.');
  }

  const normalized = validateSession(session);
  const database = await getDatabase();
  const profileId = await getCurrentProfileId();

  await database.execute(
    `UPDATE study_sessions SET
      start_at = ?, end_at = ?, duration_minutes = ?, subject_id = ?, subject_name = ?, topic = ?,
      chapter_tag = ?, mood = ?, notes = ?, mode = ?, break_minutes = ?, updated_at = CURRENT_TIMESTAMP
     WHERE id = ? AND profile_id = ?;`,
    [
      normalized.startAt,
      normalized.endAt,
      normalized.durationMinutes,
      normalized.subjectId ?? null,
      normalized.subjectName ?? null,
      normalized.topic ?? null,
      normalized.chapterTag ?? null,
      normalized.mood ?? null,
      normalized.notes ?? null,
      normalized.mode,
      normalized.breakMinutes ?? 0,
      normalized.id as number,
      profileId
    ]
  );
  syncTrigger.notifyChange();
}

export async function deleteSession(id: number): Promise<void> {
  const database = await getDatabase();
  const profileId = await getCurrentProfileId();
  await database.execute('UPDATE study_sessions SET deleted_at = CURRENT_TIMESTAMP, updated_at = CURRENT_TIMESTAMP WHERE id = ? AND profile_id = ?;', [id, profileId]);
  syncTrigger.notifyChange();
}

export async function getSessionById(id: number): Promise<StudySession | null> {
  const database = await getDatabase();
  const profileId = await getCurrentProfileId();
  const rows = await database.select<SessionRow[]>('SELECT * FROM study_sessions WHERE deleted_at IS NULL AND id = ? AND profile_id = ? LIMIT 1;', [id, profileId]);
  return rows[0] ? mapSession(rows[0]) : null;
}

/** Profile-scoped list (notifications + cross-profile reads use this). */
export async function listSessionsForProfile(profileId: number, filter: SessionFilter = {}): Promise<StudySession[]> {
  const database = await getDatabase();
  const where: string[] = ['deleted_at IS NULL'];
  const params: Array<string | number> = [];

  const limit = filter.limit ?? 50;
  const offset = filter.offset ?? 0;

  where.push('profile_id = ?');
  params.push(profileId);

  if (filter.subjectId !== undefined) {
    where.push('subject_id = ?');
    params.push(filter.subjectId);
  }

  if (filter.mode) {
    where.push('mode = ?');
    params.push(filter.mode);
  }

  if (filter.startFrom) {
    where.push('start_at >= ?');
    params.push(filter.startFrom);
  }

  if (filter.startTo) {
    where.push('start_at <= ?');
    params.push(filter.startTo);
  }

  if (filter.query) {
    where.push('(subject_name LIKE ? OR topic LIKE ? OR notes LIKE ?)');
    const query = `%${normalizeText(filter.query)}%`;
    params.push(query, query, query);
  }

  const sql = `SELECT * FROM study_sessions ${where.length ? `WHERE ${where.join(' AND ')}` : ''} ORDER BY start_at DESC LIMIT ? OFFSET ?;`;
  params.push(limit, offset);

  const rows = await database.select<SessionRow[]>(sql, params);
  return rows.map(mapSession);
}

export async function listSessions(filter: SessionFilter = {}): Promise<StudySession[]> {
  const profileId = await getCurrentProfileId();
  return listSessionsForProfile(profileId, filter);
}

export async function getRecentSessions(limit = 5): Promise<StudySession[]> {
  return listSessions({ limit });
}

export async function getSessionSummary(): Promise<SessionSummary> {
  const database = await getDatabase();
  const profileId = await getCurrentProfileId();
  const totals = await database.select<Array<{ total_sessions: number; total_minutes: number; average_minutes: number }>>(
    'SELECT COUNT(*) AS total_sessions, COALESCE(SUM(duration_minutes), 0) AS total_minutes, COALESCE(AVG(duration_minutes), 0) AS average_minutes FROM study_sessions WHERE deleted_at IS NULL AND profile_id = ?;',
    [profileId]
  );
  const recentSessions = await getRecentSessions(5);

  return {
    totalSessions: totals[0]?.total_sessions ?? 0,
    totalMinutes: totals[0]?.total_minutes ?? 0,
    averageMinutes: totals[0]?.average_minutes ?? 0,
    recentSessions
  };
}

/** Used by notifications: any non-deleted session starting on the same local calendar day as `when`. */
export async function hasSessionOnLocalCalendarDay(profileId: number, when: Date): Promise<boolean> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ c: number }>>(
    `SELECT COUNT(*) AS c FROM study_sessions
     WHERE profile_id = ?
       AND deleted_at IS NULL
       AND date(start_at, 'localtime') = date(?, 'localtime')`,
    [profileId, when.toISOString()]
  );
  return (rows[0]?.c ?? 0) > 0;
}

export async function getLastSessionStartAtIso(profileId: number): Promise<string | null> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ start_at: string }>>(
    `SELECT start_at FROM study_sessions
     WHERE profile_id = ? AND deleted_at IS NULL
     ORDER BY start_at DESC LIMIT 1`,
    [profileId]
  );
  return rows[0]?.start_at ?? null;
}

export async function sumDurationMinutesBetween(
  profileId: number,
  startIsoInclusive: string,
  endIsoInclusive: string
): Promise<number> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ m: number }>>(
    `SELECT COALESCE(SUM(duration_minutes), 0) AS m FROM study_sessions
     WHERE profile_id = ?
       AND deleted_at IS NULL
       AND datetime(start_at) >= datetime(?)
       AND datetime(start_at) <= datetime(?)`,
    [profileId, startIsoInclusive, endIsoInclusive]
  );
  return rows[0]?.m ?? 0;
}

export async function sessionCountBetween(
  profileId: number,
  startIsoInclusive: string,
  endIsoInclusive: string
): Promise<number> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ c: number }>>(
    `SELECT COUNT(*) AS c FROM study_sessions
     WHERE profile_id = ?
       AND deleted_at IS NULL
       AND datetime(start_at) >= datetime(?)
       AND datetime(start_at) <= datetime(?)`,
    [profileId, startIsoInclusive, endIsoInclusive]
  );
  return rows[0]?.c ?? 0;
}

/** Most-studied display name last `daysBack` sliding window (excluding empty names → "study"). */
export async function getDominantSubjectLastDays(profileId: number, daysBack: number): Promise<{ name: string; count: number } | null> {
  const db = await getDatabase();
  const now = new Date();
  const since = new Date(now.getFullYear(), now.getMonth(), now.getDate());
  since.setDate(since.getDate() - daysBack);
  const rows = await db.select<Array<{ sn: string; c: number }>>(
    `SELECT COALESCE(NULLIF(trim(subject_name), ''), 'General') AS sn, COUNT(*) AS c
     FROM study_sessions
     WHERE profile_id = ?
       AND deleted_at IS NULL
       AND datetime(start_at) >= datetime(?)
     GROUP BY sn
     ORDER BY c DESC
     LIMIT 1`,
    [profileId, since.toISOString()]
  );
  const r = rows[0];
  if (!r || !r.sn) return null;
  return { name: r.sn, count: r.c };
}

/** Sorted local calendar dates 'YYYY-MM-DD' with ≥1 minute logged (notifications / streak aggregation). */
export async function distinctStudyLocalDayKeysAscending(profileId: number): Promise<string[]> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ d: string }>>(
    `SELECT DISTINCT date(start_at, 'localtime') AS d
     FROM study_sessions
     WHERE profile_id = ? AND deleted_at IS NULL AND duration_minutes > 0
     ORDER BY d ASC`,
    [profileId]
  );
  return rows.map((x) => x.d);
}

/** Highest subject by minutes in `[start,end]` range. */
export async function topSubjectMinutesInRange(
  profileId: number,
  startIsoInclusive: string,
  endIsoInclusive: string
): Promise<{ name: string; minutes: number } | null> {
  const database = await getDatabase();
  const rows = await database.select<Array<{ sn: string; m: number }>>(
    `SELECT COALESCE(NULLIF(trim(subject_name), ''), 'Sessions') AS sn,
            SUM(duration_minutes) AS m
     FROM study_sessions
     WHERE profile_id = ?
       AND deleted_at IS NULL
       AND datetime(start_at) >= datetime(?)
       AND datetime(start_at) <= datetime(?)
     GROUP BY sn
     ORDER BY m DESC
     LIMIT 1`,
    [profileId, startIsoInclusive, endIsoInclusive]
  );
  const r = rows[0];
  if (!r) return null;
  return { name: r.sn, minutes: r.m };
}

function mapSession(row: any): StudySession {
  const sessionRow = row as SessionRow;
  return {
    id: sessionRow.id,
    startAt: sessionRow.start_at,
    endAt: sessionRow.end_at,
    durationMinutes: sessionRow.duration_minutes,
    subjectId: sessionRow.subject_id,
    subjectName: sessionRow.subject_name,
    topic: sessionRow.topic,
    chapterTag: sessionRow.chapter_tag,
    mood: sessionRow.mood,
    notes: sessionRow.notes,
    mode: sessionRow.mode,
    breakMinutes: sessionRow.break_minutes,
    createdAt: sessionRow.created_at,
    updatedAt: sessionRow.updated_at
  };
}
